File: .skills/openclaw-skills/skills/chyern/multi-step-workflow/scripts/task-tracker.py
#!/usr/bin/env python3
"""task-tracker - Lightweight multi-step task tracker

Usage:
  task-tracker.py new "<task>" "<step1|step2|step3>"
  task-tracker.py done "<task>" <step_number>
  task-tracker.py status "<task>"
  task-tracker.py list
  task-tracker.py clear "<task>"

Task data is stored in: ~/.openclaw/workspace/project/task-tracker/
(kept in workspace, not inside the skill package)
"""
import json, os, sys, datetime

TRACKER_DIR = os.path.expanduser("~/.openclaw/workspace/project/task-tracker")
os.makedirs(TRACKER_DIR, exist_ok=True)

def task_id(name):
    # Use hex encoding for safe filenames, preserving uniqueness
    return name.encode('utf-8').hex()

def load_task(name):
    path = os.path.join(TRACKER_DIR, f"{task_id(name)}.json")
    if not os.path.exists(path):
        return None
    with open(path) as f:
        return json.load(f)

def save_task(name, data):
    path = os.path.join(TRACKER_DIR, f"{task_id(name)}.json")
    with open(path, "w") as f:
        json.dump(data, f, indent=2, ensure_ascii=False)

def cmd_new(name, steps_raw):
    steps = [s.strip() for s in steps_raw.split('|') if s.strip()]
    data = {
        "name": name,
        "status": "in_progress",
        "steps": steps,
        "done": [],
        "created": datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    }
    save_task(name, data)
    print(f"[OK] Task created: {name} ({len(steps)} steps)")
    return data

def cmd_done(name, step_num):
    data = load_task(name)
    if not data:
        print(f"[ERROR] Task not found: {name}")
        return
    step_idx = int(step_num) - 1
    if step_idx not in data["done"]:
        data["done"].append(step_idx)
    data["done"].sort()
    if len(data["done"]) == len(data["steps"]):
        data["status"] = "completed"
    save_task(name, data)
    remaining = len(data["steps"]) - len(data["done"])
    if remaining == 0:
        print(f"[DONE] All steps completed!")
    else:
        print(f"[OK] Step {step_num} done. {remaining} remaining.")

def cmd_status(name):
    data = load_task(name)
    if not data:
        print(f"[ERROR] Task not found: {name}")
        return
    print(f"Task: {data['name']}")
    print(f"Status: {data['status']}  Progress: {len(data['done'])}/{len(data['steps'])}")
    print()
    for i, s in enumerate(data["steps"]):
        mark = "[x]" if i in data["done"] else "[ ]"
        print(f"  {mark} {i+1}. {s}")

def cmd_list():
    files = [f for f in os.listdir(TRACKER_DIR) if f.endswith('.json')]
    if not files:
        print("[INFO] No active tasks.")
        return
    print(f"[TASKS] {len(files)} active:")
    for f in sorted(files):
        with open(os.path.join(TRACKER_DIR, f)) as fp:
            d = json.load(fp)
        bar = "#" * len(d["done"]) + "-" * (len(d["steps"]) - len(d["done"]))
        print(f"  {d['name']} [{bar}] {len(d['done'])}/{len(d['steps'])}")

def cmd_clear(name):
    path = os.path.join(TRACKER_DIR, f"{task_id(name)}.json")
    if os.path.exists(path):
        os.remove(path)
        print(f"[DELETED] {name}")
    else:
        print(f"[ERROR] Task not found: {name}")

if __name__ == "__main__":
    cmd = sys.argv[1] if len(sys.argv) > 1 else ""
    if cmd == "new" and len(sys.argv) >= 4:
        cmd_new(sys.argv[2], sys.argv[3])
    elif cmd == "done" and len(sys.argv) >= 4:
        cmd_done(sys.argv[2], sys.argv[3])
    elif cmd == "status" and len(sys.argv) >= 3:
        cmd_status(sys.argv[2])
    elif cmd == "list":
        cmd_list()
    elif cmd == "clear" and len(sys.argv) >= 3:
        cmd_clear(sys.argv[2])
    else:
        print("Usage:")
        print("  task-tracker.py new \"<task>\" \"<step1|step2|step3>\"")
        print("  task-tracker.py done \"<task>\" <step_number>")
        print("  task-tracker.py status \"<task>\"")
        print("  task-tracker.py list")
        print("  task-tracker.py clear \"<task>\"")

File: .skills/openclaw-skills/skills/chyern/multi-step-workflow/scripts/task-tracker.js
#!/usr/bin/env node
/**
 * task-tracker — Track steps for complex tasks
 * 
 * Usage:
 *   node task-tracker.js new  "<task>" "<step1|step2|step3>"
 *   node task-tracker.js done "<task>" <step_number>
 *   node task-tracker.js list
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync, readdirSync, chmodSync } from 'fs';
import { resolve, join } from 'path';

import { getTempDir } from './path-resolver.js';

const TRACKER_DIR = join(getTempDir(), 'tasks');
if (!existsSync(TRACKER_DIR)) {
  mkdirSync(TRACKER_DIR, { recursive: true });
}

function taskFile(name) {
  return join(TRACKER_DIR, Buffer.from(name).toString('hex') + '.json');
}

function load(name) {
  const f = taskFile(name);
  if (!existsSync(f)) return null;
  return JSON.parse(readFileSync(f, 'utf8'));
}

function save(name, data) {
  const f = taskFile(name);
  writeFileSync(f, JSON.stringify(data, null, 2));
  try {
    chmodSync(f, 0o600);
  } catch (e) {
    // Failsafe
  }
}

const [cmd, arg1, arg2] = process.argv.slice(2);

if (!cmd) {
  console.log('Commands: new, done, list');
  process.exit(0);
}

if (cmd === 'new') {
  if (!arg1 || !arg2) { console.log('Usage: new "<task>" "<step1|step2|...>"'); process.exit(1); }
  const steps = arg2.split('|').map(s => s.trim()).filter(Boolean);
  save(arg1, { task: arg1, steps, done: [], createdAt: new Date().toISOString() });
  console.log(`[OK] Created "${arg1}" with ${steps.length} steps.`);
  steps.forEach((s, i) => console.log(`  ⏳ ${i + 1}. ${s}`));
}
else if (cmd === 'done') {
  if (!arg1 || !arg2) { console.log('Usage: done "<task>" <step_number>'); process.exit(1); }
  const data = load(arg1);
  if (!data) { console.log(`Task "${arg1}" not found.`); process.exit(1); }
  const idx = parseInt(arg2, 10) - 1;
  if (idx < 0 || idx >= data.steps.length) { console.log(`Invalid step number.`); process.exit(1); }
  if (!data.done.includes(idx)) data.done.push(idx);
  save(arg1, data);
  const remaining = data.steps.length - data.done.length;
  console.log(`[OK] Step ${idx + 1} done. ${remaining} remaining.`);
  if (remaining === 0) console.log(`[DONE] All steps completed!`);
}
else if (cmd === 'list') {
  const files = readdirSync(TRACKER_DIR).filter(f => f.endsWith('.json'));
  if (files.length === 0) { console.log('No tracked tasks.'); process.exit(0); }
  files.forEach(f => {
    const data = JSON.parse(readFileSync(join(TRACKER_DIR, f), 'utf8'));
    const progress = `${data.done.length}/${data.steps.length}`;
    console.log(`\n📋 ${data.task} [${progress}]`);
    data.steps.forEach((s, i) => {
      const mark = data.done.includes(i) ? '✅' : '⏳';
      console.log(`  ${mark} ${i + 1}. ${s}`);
    });
  });
}
else {
  console.log(`Unknown command: ${cmd}`);
  process.exit(1);
}
